packages/core/src/component.ts

```diff
@@ -3,34 +3,52 @@
 // This file is licensed under the MIT License.
 // License text available at https://opensource.org/licenses/MIT
 
-import {Constructor, Provider, BoundValue} from '@loopback/context';
-import {Application} from '.';
-
-// tslint:disable:no-any
+import { Constructor, Provider, BoundValue, Context } from '@loopback/context';
+import {Server} from '.';
 
 export interface ProviderMap {
   [key: string]: Constructor<Provider<BoundValue>>;
 }
 
+export interface ControllerMap {
+  [key: string]: Constructor<{}>;
+}
+
 export interface Component {
-  controllers?: Constructor<any>[];
+  controllers?: ControllerMap;
   providers?: ProviderMap;
+
+  // tslint:disable-next-line:no-any
   [prop: string]: any;
 }
 
+/**
+ * Mount a component to a Context. Examples can include mounting components
+ * to Server instances to attach server-specific components (ex. adding
+ * authentication to a RestServer instance).
+ *
+ * Components can also be mounted to Application instances to provide app-wide
+ * access to components. (ex. adding a messaging service to allow a RestServer
+ * to communicate with another server, like gRPC)
+ *
+ * @export
+ * @param {Context} context
+ * @param {Component} component
+ */
 export function mountComponent(
-  app: Application,
+  context: Context,
   component: Component,
 ) {
+  // TODO: Incorporate namespacing into this?
   if (component.controllers) {
-    for (const controllerCtor of component.controllers) {
-      app.controller(controllerCtor);
+    for (const controllerKey in component.controllers) {
+      context.bind(controllerKey).toClass(component.controllers[controllerKey]);
```

Comment:
❗️ 
IIUC, this is significantly changing the API contract, violating the design we agreed on in wiki and IMO adding more friction to app development 

before:

```ts
this.controllers = [ProductController, CategoryController];
```

after:

```ts
this.controllers = {
  'controllers.ProductController': ProductController,
  'controllers.CategoryController': CategoryController,
};
```

It's also shifting the responsibility of using the correct namespace in the binding key down to all app developers, opening doors to mistakes.

I personally like the current ("before") version more, although I agree your design is more flexible because we can have two controller classes with the same name mapped to different binding keys. IMO we need to carefully consider the implications and tradeoffs before making this change.